TLS sequences are not understood by SPARC/Solaris linker. GCC generates the TLS
sequence in a different order than the TLS document suggests. GNU linker can
handle this, however we are using Solaris linker, which doesn't understand this
sequence, and hence generated TLS code is incorrect.

This was reported to upstream, fixed there, and now backported into GCC 7.

Upstream issue:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=93704

--- gcc-7.3.0/gcc/config/sparc/sparc.c
+++ gcc-7.3.0/gcc/config/sparc/sparc.c
@@ -3899,11 +3899,8 @@ eligible_for_call_delay (rtx_insn *trial
   if (get_attr_in_branch_delay (trial) == IN_BRANCH_DELAY_FALSE)
     return 0;
 
-  /* Binutils allows
-       call __tls_get_addr, %tgd_call (foo)
-        add %l7, %o0, %o0, %tgd_add (foo)
-     while Sun as/ld does not.  */
-  if (TARGET_GNU_TLS || !TARGET_TLS)
+  /* The only problematic cases are TLS sequences with Sun as/ld.  */
+  if ((TARGET_GNU_TLS && HAVE_GNU_LD) || !TARGET_TLS)
     return 1;
 
   pat = PATTERN (trial);
